#1693: Refactor versions listing for packages

jbertran opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: We're adding a new OS, so we need OS-aware versions definition.

**Summary**:

The goal is to provide an authoritative listing of required packages that it as easy as possible to manage.

```
# Common packages are packages for which we need not care about OS-specific
# divergences.
#
# In this case, either:
#   * the _latest_ version is good enough, and will be the one
#     selected by the package managers (so far: apt and yum).
#   * we have strict version requirements that span OS families, and the
#     versioning schemes _and_ package names do not diverge
#
# Strict version requirements are notably:
#   * kubelet and kubectl which _make_ the K8s version of the cluster
#   * salt-minion which _makes_ the Salt version of the cluster
#
# These common packages may be overridden by OS-specific packages if package
# names or versioning conventions diverge.
#
# Packages that we build ourselves require a version and release as part of
# their build process.
```

**Acceptance criteria**: tests pass, debian files are generated with the correct versions.

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1693

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

jbertran committed:
packaging: refactor download list generation

jbertran committed:
packaging: refactor build list generation

jbertran committed:
packaging: reorder definitions and add folding

jbertran committed:
versions: introduce a PackageVersion class

This allows us to control package listings more finely, and also
provides helpers for version and fully versioned package name
serialization to strings.

jbertran committed:
versions: introduce common packages and extras/overrides

The support of an additional OS requires a 'compatibility' layer for
package names and versions.

We define a common core of packages whose names do not diverge across OS
families and whose versions are either latest or fixed across OS families.

We introduce overrides in package versions, such that a package for a specific
version may override the package referenced in the common core.

We unpin most packages in the RPM listing, where we only really need to pin
K8s core components as well as some of the 'unstable' Python APIs we rely
on.

bert-e commented:
# Hello jbertran,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

slaperche-scality commented on buildchain/buildchain/versions.py:
```diff
@@ -363,6 +363,106 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
 PACKAGES_MAP = {pkg.name: pkg for pkg in RPM_PACKAGES}
 
 
+DEB_PACKAGES = (
```
This commit should probably be removed (`DEB_PACKAGES` is already defined l.466)

jbertran replied:
Oops, indeed. Rebase gone wrong

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -324,18 +324,27 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
-    if pkg.name not in _RPM_TO_BUILD_PKG_NAMES
-}
+_TO_DOWNLOAD_RPM_CONFIG : Dict[str, str] = _list_packages_to_download(
+    versions.PACKAGES,
```
minor: can be wrapped:

```python
_TO_DOWNLOAD_RPM_CONFIG : Dict[str, str] = _list_packages_to_download(
    versions.PACKAGES, _RPM_TO_BUILD_PKG_NAMES
)
```

ditto for `_TO_DOWNLOAD_DEB_CONFIG`

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -324,18 +324,27 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
-    for pkg in versions.DEB_PACKAGES
-    if pkg.name not in _DEB_TO_BUILD_PKG_NAMES
-}
+# Store these versions in a dict to use with doit.tools.config_changed
```
Not sure we want to duplicate this comment

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -308,12 +308,32 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
+    pkg_cats: Dict[str, Tuple[targets.Package, ...]]
+) -> List[str]:
+    return [
+        pkg.name for pkg_list in pkg_cats.values()
```
Can be squashed:

```python
    return [
        pkg.name for pkg_list in pkg_cats.values() for pkg in pkg_list
    ]
```

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -202,6 +227,12 @@ def task__build_rpm_repositories() -> Iterator[types.TaskDict]:
 
+
+# }}}
+
```
Remove empty lines between folds

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -239,6 +270,10 @@ def task__build_rpm_repositories() -> Iterator[types.TaskDict]:
 )
 
+# }}}
+
```
ditto

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -381,6 +349,65 @@ def _list_packages_to_download(
 )
 
+# }}}
+
```
ditto

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -328,8 +327,8 @@ def _rpm_package(name: str, sources: List[Path]) -> targets.RPMPackage:
-    "{p.name}-{p.version}-{p.release}".format(p=package)
-    for package in versions.RPM_PACKAGES
+    package.rpm_full_name
+    for package in versions.PACKAGES
```
Why is `RPM_PACKAGES` renamed `PACKAGES`?

jbertran replied:
Gah, stray rebase conflict resolution, thank you

slaperche-scality commented on buildchain/buildchain/versions.py:
```diff
@@ -364,200 +425,65 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
 
 DEB_PACKAGES = (
-    Package(
+    PackageVersion(
```
Save some screen real estate and squash into a single line: `PackageVersion(name='containerd'),` (ditto for the others)

jbertran replied:
Done in a following commit

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

jbertran committed:
versions: switch to `Mapping` type for built packages return value

The `Dict` type is too strict and causes mypy to expect Package type
where we should expect Package _or_ any subclass of Package type.

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -1,6 +1,4 @@
 # coding: utf-8
-
```
Why?

Also, it become inconsistent with the other files.

slaperche-scality commented on buildchain/buildchain/versions.py:
```diff
@@ -228,132 +228,193 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
+        self._version = version
+        self._release = release
+
+    name = property(operator.attrgetter('_name'))
```
minor: would align

```
    name    = property(operator.attrgetter('_name'))
    version = property(operator.attrgetter('_version'))
    release = property(operator.attrgetter('_release'))
```

jbertran replied:
Damn that triggers my linter like crazy. Done.

slaperche-scality commented on buildchain/buildchain/versions.py:
```diff
@@ -228,132 +228,193 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
+    @property
+    def rpm_full_name(self) -> str:
+        """The package's full name in RPM conventions."""
+        name_fmt = '{p.name}'
```
could it be rewritten:

```python
if self.version:
    return '{}-{}'.format(self.name, self.full_version)
return self.name
```
?

slaperche-scality commented on buildchain/buildchain/versions.py:
```diff
@@ -228,132 +228,193 @@ def _version_prefix(version: str, prefix: str = 'v') -> str:
+        return name_fmt.format(p=self)
+
+    @property
+    def deb_full_name(self) -> str:
```
ditto, could we do something like:

```
if self.version:
    return '{}={}'.format(self.name, self.full_version)
return self.name
```
?

jbertran replied:
Oh indeed, that's basic DRY. Thanks.

slaperche-scality commented on buildchain/buildchain/versions.py:
```diff
@@ -291,201 +294,113 @@ def deb_full_name(self) -> str:
+    os_family_pkgs = PACKAGES.get(os_family)
+
+    if os_family_pkgs is None:
+        raise Exception('Unlisted OS family: {}'.format(os_family))
```
s/Unlisted/Unsupported/?

jbertran replied:
Eeeeh technically we don't support Ubuntu yet but we download its packages. As `versions` is supposed to be absolutely standalone I'd rather observe things that pertain to it and only to it, hence unlisted in the context of the package dict keys.

slaperche-scality replied:
Make sense, but I think we can have a better word than `unlisted`.
Maybe `no packages for OS family {}`?

jbertran replied:
Suits me, thanks

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -52,7 +52,7 @@ def _list_packages_to_build(
     package_versions: Tuple[versions.PackageVersion, ...],
     packages_to_build: List[str]
-) -> Dict[str,str]:
+) -> Mapping[str, Optional[str]]:
```
Why the change here? (and BTW, why the change in general: a commit message would be welcome).

IIRC, the good practice is to use "precise" type for return type and "abstract" one for input.

See https://docs.python.org/3/library/typing.html#typing.Dict:

> Useful for annotating return types. To annotate arguments it is preferred to use an abstract collection type such as Mapping.

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -346,6 +339,61 @@ def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
+
+def _deb_package(name: str, sources: Path) -> targets.DEBPackage:
+    try:
+        pkg_info = versions.RPM_PACKAGES_MAP[name]
```
`_deb_package` using `RPM_PACKAGES_MAP`?

jbertran replied:
Oh boy that's bad.

slaperche-scality commented on buildchain/buildchain/packaging.py:
```diff
@@ -27,7 +27,7 @@
 
 from pathlib import Path
-from typing import Dict, FrozenSet, Iterator, List, Tuple
+from typing import Dict, FrozenSet, Iterator, List, Mapping, Optional, Tuple
```
Still no explanation on why this switch in the commit message

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

jbertran commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye jbertran.
